src/libraries/Common/src/System/Globalization/FormatProvider.Number.cs

```diff
@@ -517,15 +517,19 @@ private static unsafe bool ParseNumber(ref char* str, char* strEnd, NumberStyles
                             {
                                 exp = exp * 10 + (ch - '0');
                                 ch = ++p < strEnd ? *p : '\0';
-                                if (exp > 1000)
-                                {
-                                    exp = 9999;
-                                    while (ch >= '0' && ch <= '9')
-                                    {
-                                        ch = ++p < strEnd ? *p : '\0';
-                                    }
-                                }
-                            } while (ch >= '0' && ch <= '9');
+                            } while ((ch >= '0') && (ch <= '9') && (exp < int.MaxValue / 10));
+
+                            if ((ch >= '0') && (ch <= '9'))
+                            {
+                                // We still had remaining characters but bailed early because
+                                // the exponent was going to overflow. If exp is exactly 214748364
+                                // then we can technically handle one more character being 0-9
```

Comment:
```suggestion
                                // then we can technically handle one more character being 0-7
```

`2147483647` is `int.MaxValue` so the last digit, if we handled it, could only be `0-7`